### DISCUSSION TITLE
Is there a way to access to viewmodel from MainWindow.cs?

### DISCUSSION THREAD

**Discussion by btylmz** [MvvmCross/MvvmCross#4447] [Q&A] [answered]
I am currently developing an app on WPF platform with a serial port connection. Currently using the MVVMCross 8.0.2 version. In the app a receiver thread always listens the serial port for incoming data inside an infinite loop. When the application is closed thread continous to work and application doesnt stop. 

I am trying to solve this problem for a while, what i've tried so far is using the view appear/disappear functions, however, these functions are not called when the app is closed by the window close button.

So, i thought if i can access to viewmodels from the mainwindow.cs, i can stop the thread using the window_closing event. Is there a way to do that?

Or if you have any different idea to stop the thread when the app is closed, i would appreciate the help.

**Comment by Cheesebaron** [comment, marked as answer]
Sounds like you are doing some weird stuff in your ViewModel and it has the responsibility for listening to your serial port socket. This should probably be offloaded to a service layer in your App instead. Then you could resolve this service layer part using the IoC.

**Reply by btylmz** [reply]
Thanks for your response, sorry i was a bit busy couldn't respond to your answer. I am a junior developer and i can't understand how a service layer would solve my problem. I am still learning design patterns and topics such as IoC, dependency injection etc. If i implement a service layer in the core project wouldn't i get the same problem?  What i want to know is how to access to the viewmodels or models from the mainwindow.cs. If a service layer would service this problem can you give more detail about how to implement it?

**Edit**

I solved the problem by breaking the MVVM structure. It is helpful for initial tests at least but I want to find a better solution. I added the window closing event behind the view and by accessing the datacontext i am stopping the thread.

**Reply by entdark** [reply]
Join https://aka.ms/dotnet-discord and ask your question there. It is unrelated to MvvmCross at all.
